internal/jujuapi/streamproxy.go

```diff
@@ -0,0 +1,150 @@
+	email := jwtToken.Subject()
+	ctx = auth.ContextWithSessionIdentity(ctx, email)
+	return ctx, nil
+}
+
+// ServeWS implements jimmhttp.WSServer.
+func (s streamProxier) ServeWS(ctx context.Context, clientConn *websocket.Conn) {
+	writeError := func(msg string, code errors.Code) {
+		var errResult jujuparams.ErrorResult
+		errResult.Error = &jujuparams.Error{
+			Message: msg,
+			Code:    string(code),
+		}
+		err := clientConn.WriteJSON(errResult)
+		if err != nil {
+			zapctx.Error(ctx, "failed to write error message to client", zap.Error(err), zap.Any("client message", errResult))
+		}
+	}
+	user, err := s.jimm.UserLogin(ctx, auth.SessionIdentityFromContext(ctx))
+	if err != nil {
+		zapctx.Error(ctx, "user login error", zap.Error(err))
+		writeError(err.Error(), errors.CodeUnauthorized)
+		return
+	}
+	uuid, finalPath, err := modelInfoFromPath(jimmhttp.PathElementFromContext(ctx, "path"))
+	if err != nil {
+		zapctx.Error(ctx, "error parsing path", zap.Error(err))
+		writeError(fmt.Sprintf("error parsing path: %s", err.Error()), errors.CodeBadRequest)
+		return
+	}
+	model, err := s.getModel(ctx, uuid)
+	if err != nil {
+		writeError(err.Error(), errors.CodeModelNotFound)
+		return
+	}
+	if ok, err := checkPermission(ctx, finalPath, user, model.ResourceTag()); err != nil {
+		writeError(err.Error(), errors.CodeUnauthorized)
+		return
+	} else if !ok {
+		writeError(fmt.Sprintf("unauthorized access to endpoint: %s", finalPath), errors.CodeUnauthorized)
+		return
+	}
+	api, err := s.jimm.Dialer.Dial(ctx, &model.Controller, model.ResourceTag(), nil)
+	if err != nil {
+		zapctx.Error(ctx, "failed to dial controller", zap.Error(err))
+		writeError(fmt.Sprintf("failed to dial controller: %s", err.Error()), errors.CodeConnectionFailed)
+		return
+	}
+	defer api.Close()
+	controllerStream, err := api.ConnectStream(finalPath, nil)
+	if err != nil {
+		zapctx.Error(ctx, "failed to connect stream", zap.Error(err))
+		writeError(fmt.Sprintf("failed to connect stream: %s", err.Error()), errors.CodeConnectionFailed)
+		return
+	}
+	proxyStreams(clientConn, controllerStream)
+}
+
+func (s streamProxier) getModel(ctx context.Context, modelUUID string) (dbmodel.Model, error) {
+	model := dbmodel.Model{
+		UUID: sql.NullString{
+			String: modelUUID,
+			Valid:  modelUUID != "",
+		},
+	}
+	if err := s.jimm.Database.GetModel(context.Background(), &model); err != nil {
+		zapctx.Error(ctx, "failed to find model", zap.String("uuid", modelUUID), zap.Error(err))
+		return dbmodel.Model{}, fmt.Errorf("failed to find model: %s", err.Error())
+	}
+	return model, nil
+}
+
+// proxyStreams starts a simple proxy for 2 websockets.
+// After starting the proxy we listen for the first error
+// returned and then close both connections before waiting
+// for an error from the second connection.
+func proxyStreams(src, dst base.Stream) {
+	errChan := make(chan error, 2)
+	go func() { errChan <- proxy(src, dst) }()
+	go func() { errChan <- proxy(dst, src) }()
+	<-errChan
```

Comment:
shouldn't we catch and log these errors?

Reply:
I'm inclined to say no because we can't do anything with these errors besides possibly logging them.

Reply:
i do think logging the error in this case is valid.
Just imagine the stream proxy not working without any error logs.

Reply:
Adding logging for them.